extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/propagation/PassThroughPropagator.java

```diff
@@ -0,0 +1,99 @@
+import java.util.stream.StreamSupport;
+import javax.annotation.Nullable;
+
+/**
+ * A {@link TextMapPropagator} which can be configured with a set of fields, which will be extracted
+ * and stored in {@link Context}. If the {@link Context} is used again to inject, the values will be
+ * injected as-is. This {@link TextMapPropagator} is appropriate for a service that does not need to
+ * participate in telemetry in any way and provides the most efficient way of propagating incoming
+ * context to outgoing requests. In almost all cases, you will configure this single {@link
+ * TextMapPropagator} when using {@link
+ * io.opentelemetry.api.OpenTelemetry#propagating(ContextPropagators)} to create an {@link
+ * io.opentelemetry.api.OpenTelemetry} that only propagates. Similarly, you will never need this
+ * when using the OpenTelemetry SDK to enable telemetry.
+ */
+public class PassThroughPropagator implements TextMapPropagator {
+
+  private static final ContextKey<List<String>> EXTRACTED_KEY_VALUES =
+      ContextKey.named("passthroughpropagator-keyvalues");
+
+  /**
+   * Returns a {@link PassThroughPropagator} which will propagate the given {@code fields} from
+   * extraction to injection.
+   */
+  public static PassThroughPropagator create(String... fields) {
+    requireNonNull(fields, "fields");
+    return create(Arrays.asList(fields));
+  }
+
+  /**
+   * Returns a {@link PassThroughPropagator} which will propagate the given {@code fields} from
+   * extraction to injection.
+   */
+  public static PassThroughPropagator create(Iterable<String> fields) {
+    requireNonNull(fields, "fields");
+    List<String> fieldsList =
+        StreamSupport.stream(fields.spliterator(), false)
+            .map(field -> requireNonNull(field, "field"))
+            .collect(Collectors.toList());
+    return new PassThroughPropagator(fieldsList);
+  }
+
+  private final List<String> fields;
+
+  private PassThroughPropagator(List<String> fields) {
+    this.fields = Collections.unmodifiableList(fields);
+  }
+
+  @Override
+  public Collection<String> fields() {
+    return fields;
+  }
+
+  @Override
+  public <C> void inject(Context context, @Nullable C carrier, TextMapSetter<C> setter) {
+    List<String> extracted = context.get(EXTRACTED_KEY_VALUES);
+    if (extracted != null) {
+      for (int i = 0; i < extracted.size(); i += 2) {
+        setter.set(carrier, extracted.get(i), extracted.get(i + 1));
+      }
+    }
+  }
+
+  @Override
+  public <C> Context extract(Context context, @Nullable C carrier, TextMapGetter<C> getter) {
+    List<String> extracted = null;
+    for (String field : fields) {
+      String value = getter.get(carrier, field);
+      if (value != null) {
+        if (extracted == null) {
+          extracted = new ArrayList<>();
+        }
+        extracted.add(field);
+        extracted.add(value);
+      }
+    }
+    return extracted != null ? context.with(EXTRACTED_KEY_VALUES, extracted) : context;
```

Comment:
Does context key being private provide enough safety to not wrap `extracted` with unmodifiable?

Reply:
Yes, it should be fine IMHO